Sampling from dimensions consolidated into one function

SRFU-NN opened the pull request:
* Sampling consolidated. There is now only one fundamental sampling function for a dimension: `Dimension.sample()`. It converts an index in [0, 1] to a value in the dimension. It respects priors, so `log uniform Real` has a different sampling function than `uniform Real`, and a categorical dimension with a informative prior returns categories with a higher prior more often.
* `Integer.sample()` now has non-informative priors (it has the same probability to return every integer in the `Dimension`). This partly solves issue #158 , but it masks the part of it it doesn't solve (edge points have different spaces than internal points).
* You can choose whether to get a list of unique samples, or a list of samples that is as long as the list of indices you provided.
* `Dimension.rvs()` generates a list based on a random sampling of `Dimension.sample()`, so it respects priors as before. Do we need `Dimension.rvs()`, or can we simplify it so only `Space.rvs()` exist? EDIT: `Dimension.rvs()` no longer exists. `Space.rvs()` respects the priors of the underlying dimensions.
* `Space.lhs()` samples uniformly from `Dimension.sample()`, so it now also respects priors. **THIS IS A CHANGE IN BEHAVIOR**.
* Values presented for a given seed has changed. **THIS IS A BREAKING CHANGE** for anybody who relies on seeds always giving specific results (as I do in my documentation).

Ah, crap, Github reads all of space.py as on giant deletion and insertion, instead of telling where there are changes. It is not reviewable as is. Can anybody remember how to solve that?

SRFU-NN committed:
feat: Dimenison.sample() implemented

SRFU-NN committed:
fix: Fixing sampling of integer spaces

SRFU-NN committed:
fix: INteger.sample() now returns integers

SRFU-NN committed:
feat: Dimension.rvs() via sampling approach

SRFU-NN committed:
feat: Space.lhs() uses Dimension.sample()

dk-teknologisk-mon commented:
What type of OS are you coding on? My googling hints at it being an end-of-line issue: 
https://gitlab.com/gitlab-org/gitlab-foss/-/issues/52936 https://stackoverflow.com/questions/12876350/diff-returning-entire-file-for-identical-files
https://stackoverflow.com/questions/37344280/git-diff-is-showing-full-file-has-changed-for-a-single-line-change-but-only-for
If this is the root cause, then the fix lies in tinkering with your git settings.

Edit: Yup, definitely white space/eol formatting that is the issue. In the files changed pane you can set it to ignore whitespace changes (the settings gear right under the panes), which immediately clears up the diff to show the ‘actual’ changes.

dk-teknologisk-mon commented:
After obtaining a more manageable diff, the first thing that jumps out at me is the use of RandomState, which is strongly discouraged now, see e.g. https://albertcthomas.github.io/good-practices-random-number-generators/ which explains it a bit clearer than the Numpy documentation itself. This update will totally brick scripts that rely on points generated up to this point no matter what, so we may as well fix it now?

SRFU-NN commented:
> What type of OS are you coding on? My googling hints at it being an end-of-line issue: https://gitlab.com/gitlab-org/gitlab-foss/-/issues/52936 https://stackoverflow.com/questions/12876350/diff-returning-entire-file-for-identical-files https://stackoverflow.com/questions/37344280/git-diff-is-showing-full-file-has-changed-for-a-single-line-change-but-only-for If this is the root cause, then the fix lies in tinkering with your git settings.
> 
> Edit: Yup, definitely white space/eol formatting that is the issue. In the files changed pane you can set it to ignore whitespace changes (the settings gear right under the panes), which immediately clears up the diff to show the ‘actual’ changes.

Your Google-fu is stronger than mine! But good to know there is a way around it.

SRFU-NN commented:
> After obtaining a more manageable diff, the first thing that jumps out at me is the use of RandomState, which is strongly discouraged now, see e.g. https://albertcthomas.github.io/good-practices-random-number-generators/ which explains it a bit clearer than the Numpy documentation itself. This update will totally brick scripts that rely on points generated up to this point no matter what, so we may as well fix it now?

I had it set up to use default-generator, but some of our code uses Randomstate and fails if I use the seed as input into a numpy.random.default_rng, so it is a bigger change to use the more current workflow. I can look into how big a change it would be, but I will fix what needs to be fixed first :-)

SRFU-NN commented:
> After obtaining a more manageable diff, the first thing that jumps out at me is the use of RandomState, which is strongly discouraged now, see e.g. https://albertcthomas.github.io/good-practices-random-number-generators/ which explains it a bit clearer than the Numpy documentation itself. This update will totally brick scripts that rely on points generated up to this point no matter what, so we may as well fix it now?

Actually, the easiest and most backward compatible ay would be to build an intepreter that takes in a number, a RandomState or a rng, and spits out an rng. I will look into that.

SRFU-NN commented:
We also need to check the performance of the new approach. When we do a lot of random value sampling, there could be significant performance implications.

SRFU-NN committed:
fix: Constraints work with the new rvs

SRFU-NN committed:
style: Using modern RNG, not RandomState

dk-teknologisk-mon commented:
> > After obtaining a more manageable diff, the first thing that jumps out at me is the use of RandomState, which is strongly discouraged now, see e.g. https://albertcthomas.github.io/good-practices-random-number-generators/ which explains it a bit clearer than the Numpy documentation itself. This update will totally brick scripts that rely on points generated up to this point no matter what, so we may as well fix it now?
> 
> Actually, the easiest and most backward compatible ay would be to build an intepreter that takes in a number, a RandomState or a rng, and spits out an rng. I will look into that.

If the functions still accept a RandomState when this is passed, but change the default to a seed, then I don't mind so much. In my opinion it also makes it a lot clearer where the reproducibility is ensured by explicitly building and seeding rng's where they are needed, rather than relying on the state of a global variable.

SRFU-NN committed:
test: Corrected space for log-uniform Real

dk-teknologisk-mon commented on ProcessOptimizer/space/space.py:
```diff
@@ -1,917 +1,896 @@
+                               "the space, not %s and %s." % (a, b))
+        return abs(a - b)
+
+    def _sample(self, point_list=Iterable[float]) -> np.ndarray:
```
Is the `=` for `point_list` intentional? I assume this is meant to be a type hint, in which case the line should be 
```python
def _sample(self, point_list: Iterable[float]) -> np.ndarray:
```

dk-teknologisk-mon commented on ProcessOptimizer/space/space.py:
```diff
@@ -1,917 +1,896 @@
+                               " the space, not {} and {}.".format(a, b))
+        return 1 if a != b else 0
+
+    def _sample(self, point_list=Iterable[float]) -> np.ndarray:
```
Also here, is the Iterable not meant to be a type-hint for point_list?

dk-teknologisk-mon commented on ProcessOptimizer/space/space.py:
```diff
@@ -1,917 +1,896 @@
+            # Get evenly distributed samples from one dimension
+            sample_indices = (np.arange(n) + 0.5) / n
+            lhs_aranged = self.dimensions[i].sample(sample_indices)
+            perm = np.random.RandomState(seed=42 + i).permutation(n)
```
While we are at it, I suggest changing this line to use np.random.default_rng with a set seed instead.

SRFU-NN replied:
I had completely missed that use of RandomState.

We might as well open up for actually having LHS being random, for consistency.

dk-teknologisk-mon replied:
It would certainly help benchmarking, since right now you always get the exact same LHS set of points, no matter what. This makes your results very system-specific, as you always start optimizations from the same initial points, which may happen to be really good (or bad) for the given "landscape".

SRFU-NN replied:
Fixed.

SRFU-NN committed:
test: Testing consistency of pseudorandom sampling

dk-teknologisk-mon commented on ProcessOptimizer/space/space.py:
```diff
@@ -1,917 +1,896 @@
+        * `points` [float or list[float]]:
+            A single point or a list of points to sample. All must be between 0 and 1.
+
+        * `deduplicate` [bool]:
```
A semantic suggestion here: Maybe a better word that aligns with the default behavior would be `allow_duplicates`, with a default of True? I think it would make it much easier to figure out what the key-word is controlling.

SRFU-NN replied:
Done

SRFU-NN committed:
style: Romved old Dimension.rvs() method

SRFU-NN committed:
fix: Corrected default parameters -> typehints

dk-teknologisk-mon commented on ProcessOptimizer/space/space.py:
```diff
@@ -1,917 +1,896 @@
-            Set random state to something other than None for reproducible
-            results.
-        """
-        rng = check_random_state(random_state)
```
As discussed in the comments, I'd suggest keeping the explicit assignment of an RNG here, but instead of using `check_random_state` and so on, instead change the function call so the third input is called `seed` to encourage this use, and still allow it to accept int, RandomState or None.

Then, in the generation of `index_array` you can then call `rng.uniform(size=n_samples)`, which should work no matter if the user passes a seed number or a RandomState.

SRFU-NN committed:
style: lhs uses modern RNG

SRFU-NN committed:
dc: More precise description of Dimension.distance

dk-teknologisk-mon commented on ProcessOptimizer/space/space.py:
```diff
@@ -1,917 +1,896 @@
+    ) -> np.ndarray:
+        """Draw points from the dimension.
+
+        The sampling should be a reasonable mapping from the intercal [0, 1] to the
```
Typo in "interval", and note that the implementation is on the interval [0, 1), that is 1 is not included. This was explicitly handled in the _uniform_inclusive function before (line 186), but I'm not really sure it matters with the rounding that will eventually happen when translating into specific factor settings.

SRFU-NN replied:
I have moved this description to the abstract method `_sample`, since it is an instruction for when you want to implement the abstract method.

It actually is [0,1] here.

rvs() only operates on [0,1), so it can now never give the upper bound of a real dimension. But the formal probability of that happening is 0 anyway, so I don't see that as a big issue. Do correct me if there is something I am missing.

edit: The actual probability is not 0, due to the finite nature of floats. But the probability is low enough that I don't see it as an issue.

dk-teknologisk-mon commented on ProcessOptimizer/space/space.py:
```diff
@@ -1,917 +1,922 @@
+    """
+    if input is None:
+        return np.random.default_rng()
+    elif isinstance(input, (int, float)):
```
Looking at the docs for default_rng it looks like it does not accept floats as seeds (it states `seed
{None, int, array_like[ints], SeedSequence, BitGenerator, Generator}, optional`)

SRFU-NN replied:
Yeah, I was getting ahead of myself. I have fixed it.

The way Github Copilot suggests I handle RandStates also doesn't work. Now, I use the RandomState to give me a list of integers, and used that as a seed for a default_rng.

SRFU-NN replied:
I only allow for 10 000 different values, but since the use case is to ensure reproducibility, and RandomState is deprecated, I think that should be enough.

SRFU-NN committed:
fix: RandomState now works as input

SRFU-NN committed:
feat: Changed from deduplicate to allow_duplicates

SRFU-NN committed:
doc: Moved documentation of implementation

How to implement sampling for a Dimension type is moved to the abstract
method `_sample()`

SRFU-NN committed:
fix: I don't know how that line disappeared

SRFU-NN committed:
test: np.integer is a superclass for all integers

dk-teknologisk-mon commented on ProcessOptimizer/space/space.py:
```diff
@@ -1,917 +1,942 @@
+    elif isinstance(input, int):
+        return np.random.default_rng(input)
+    elif isinstance(input, np.random.RandomState):
+        return np.random.default_rng(
```
Just to be sure I understand what is going on here: So you use whatever the global RandomState is to draw an array containing 10 integers between 0 and 1000, which you then feed to the default_rng. Why 10?

SRFU-NN replied:
The RandomState used is not necessarily the global RandomState, it is whatever RandomState the user has provided. If that has been initiated with `None` as the seed, it will be the global one, but most likely, it has been initiated with some seed - Why else create it and give it to rvs() or lhs()?

No particular reason for 10 integers in [0, 1000), just what I came up with.

Though thinking about it, it does give me significantly higher entropy than I thought - 10^30 different values, to be exact. If I am right in my assumptions about how default_rng uses its seed.

SRFU-NN commented:
A quick test indicates that my approach is around 30% slower than the current approach, meaning that space.rvs(n_samples=1000) goes from taking around 10 ms to around 13 ms. 

While I don't think this is significant, I will try and modify the solution so it uses transformers instead of math, to see if that helps. It will make the code less readable, though.

edit: the timing really isn't reproducible enough to say anything. Let's discuss at the meeting this afternoon, though I favor keeping the new code as is.

SRFU-NN committed:
test: testing values and order for lhs.

SRFU-NN committed:
test: Testinf random lhs

dk-teknologisk-mon commented on ProcessOptimizer/space/space.py:
```diff
@@ -1,917 +1,942 @@
+            raise ValueError("Sample points must be between 0 and 1.")
+        sampled_points = self._sample(points)
+        if not allow_duplicates:
+            # np.unique sorts the inputs, which we do not wan't, so we have to reinvent
```
want without '
```suggestion
            # np.unique sorts the inputs, which we do not want, so we have to reinvent
```

SRFU-NN committed:
Update ProcessOptimizer/space/space.py

Co-authored-by: Morten Bormann Nielsen <64902587+dk-teknologisk-mon@users.noreply.github.com>

SRFU-NN committed:
doc: Updatd CHANGELOG.MD

SRFU-NN committed:
style: moved get_random_generator to utils

Also refactored utils to be a folder to avoid circular import: utils.py
imports space.py, and I need get_random_generator in space.py, so I
can't define it in utils.py. By creating the utils folder, moving
utils.py into it, and importing every function in __init__.py, I keep
the functionality. I can then define get_random_generartor in
get_rng.py in the utils folder, and import that into __init__.py as
well.

SRFU-NN committed:
doc: Corrected math

SRFU-NN commented:
The pull request is ready for final review. I have added changes to CHANGELOG.md so it becomes easier to up the version.

I accidentally black formatted space.py (my editor does that on save, I just usually remember to not commit those changes). I hope that is OK, it shouldn't change any functionality. It does make `git blame` harder to use, a lot of lines are now attributed to me simply due to that.

SRFU-NN committed:
style: Changed casing in variable name

dk-teknologisk-mon commented on CHANGELOG.md:
```diff
@@ -4,7 +4,22 @@
+  `Dimesion`s
+- LHS now allows for arbitrary seeding, or for random seeding, better supporting
+  benchmarking
+- the module `space` uses `np.random.default_rng` as a random number generator, instead
```
```suggestion
- The module `space` uses `np.random.default_rng` as a random number generator, instead
```

dk-teknologisk-mon commented on CHANGELOG.md:
```diff
@@ -4,7 +4,22 @@
+  Sampling (LHS)) are handled through `Space`, referencing the sampling functions of the
+  `Dimesion`s
+- LHS now allows for arbitrary seeding, or for random seeding, better supporting
+  benchmarking
```
```suggestion
  benchmarking. The algorithm still uses a fixed seed by default.
```

dk-teknologisk-mon replied:
Make it clear that LHS still provides the same list if you call it repeatedly with default settings.

dk-teknologisk-mon commented on ProcessOptimizer/space/space.py:
```diff
@@ -1,917 +1,940 @@
+    if len(dimension) == 3:
+        if any([isinstance(dim, (float, int)) for dim in dimension[:2]]) and dimension[
+            2
+        ] in ["uniform", "log-uniform"]:
```
```suggestion
        if any([isinstance(dim, (float, int)) for dim in dimension[:2]]) and 
            dimension[2] in ["uniform", "log-uniform"]:
```
The old formatting reads better.

dk-teknologisk-mon commented on ProcessOptimizer/space/space.py:
```diff
@@ -1,917 +1,940 @@
+                if point not in seen:
+                    deduplicated_points.append(point)
+                    seen.add(point)
+            sampled_points = deduplicated_points
```
```suggestion
            unique_points = []
            for point in sampled_points:
                if point not in seen:
                    unique_points.append(point)
                    seen.add(point)
            sampled_points = unique_points
```
"deduplicated" is rather long way of writing unique

dk-teknologisk-mon commented on ProcessOptimizer/space/space.py:
```diff
@@ -1,917 +1,940 @@
+        """A reasonable mapping from the interval [0, 1] to the dimension, whatever that
+        may mean. For example, for an integer dimension, the sampling should be give a
+        higher integer for a higher value of the point, and each integer should be
+        mapped to from en equally large interval.
```
```suggestion
        mapped to from an equally large interval.
```

dk-teknologisk-mon commented on ProcessOptimizer/space/space.py:
```diff
@@ -1,917 +1,940 @@
+        orignal space.
+        """
+        return np.clip(
+            super(Real, self).inverse_transform(Xt).astype(float), self.low, self.high
```
```suggestion
            super(Real, self).inverse_transform(Xt).astype(float), 
            self.low, 
            self.high
```
Makes it easier to see that np.clip() is getting three inputs.

dk-teknologisk-mon commented on ProcessOptimizer/space/space.py:
```diff
@@ -1,917 +1,940 @@
+        if transform not in ["normalize", "identity"]:
+            raise ValueError(
+                "transform should be 'normalize' or 'identity'"
+                " got {}".format(self.transform_)
```
```suggestion
                "Transform should be 'normalize' or 'identity' "
                "got {}".format(self.transform_)
```

dk-teknologisk-mon commented on ProcessOptimizer/space/space.py:
```diff
@@ -1,917 +1,940 @@
+        if not (a in self and b in self):
+            raise RuntimeError(
+                "Can only compute distance for values within"
+                " the space, not {} and {}.".format(a, b)
```
```suggestion
                "Can only compute distance for values within "
                "the space, not {} and {}.".format(a, b)
```

dk-teknologisk-mon commented on ProcessOptimizer/space/space.py:
```diff
@@ -1,917 +1,940 @@
+        if high <= low:
+            raise ValueError(
+                "the lower bound {} has to be less than the"
+                " upper bound {}".format(low, high)
```
```suggestion
                "The lower bound {} has to be less than the "
                "upper bound {}".format(low, high)
```

dk-teknologisk-mon commented on ProcessOptimizer/space/space.py:
```diff
@@ -1,917 +1,940 @@
+        return 1 if a != b else 0
+
+    def _sample(self, point_list: Iterable[float]) -> np.ndarray:
+        # XXX check that sum(prior) == 1
```
Is this comment meant to be a future TODO, or something that was forgotten?

dk-teknologisk-mon commented on ProcessOptimizer/space/space.py:
```diff
@@ -1,917 +1,940 @@
+    def lhs(
+        self,
+        n: int,
+        random_state: Union[
```
```suggestion
        seed: Union[
```
To encourage use of seeding rather than RandomState.

dk-teknologisk-mon commented on ProcessOptimizer/space/space.py:
```diff
@@ -1,917 +1,940 @@
+        ----------
+        * `n` [int]: The number of samples to generate.
+
+        * `random_state`
```
```suggestion
        * `seed`
```

dk-teknologisk-mon commented on ProcessOptimizer/space/space.py:
```diff
@@ -1,917 +1,940 @@
+            [int, float, np.random.RandomState, np.random.Generator, or None, default=42]:
+            The seed used by the random number generator. If None, the results are not reproducible.
+        """
+        rng = get_random_generator(random_state)
```
```suggestion
        rng = get_random_generator(seed)
```

dk-teknologisk-mon commented on ProcessOptimizer/tests/test_space.py:
```diff
@@ -110,18 +84,98 @@ def test_real_bounds():
+        if i < len(deduplicated_values) - 1:
+            first_pos = np.argwhere(randomvalues == deduplicated_values[i])[0][0]
+            second_pos = np.argwhere(randomvalues == deduplicated_values[i + 1])[0][0]
+            assert first_pos < second_pos
```
```suggestion
    unique_values = dimension.sample(
        generator.random(size=50), allow_duplicates=False
    )
    assert len(unique_values) == len(set(unique_values))
    # Checking of no repeated values
    assert set(unique_values) == set(randomvalues)
    # Checking that seeding works
    for i in range(len(unique_values)):
        # Checking that order is preserved when removing duplicates
        if i < len(unique_values) - 1:
            first_pos = np.argwhere(randomvalues == unique_values[i])[0][0]
            second_pos = np.argwhere(randomvalues == unique_values[i + 1])[0][0]
            assert first_pos < second_pos
```
Again, I think it is much easier to understand "unique" instead of "deduplicated"

dk-teknologisk-mon commented on ProcessOptimizer/tests/test_space.py:
```diff
@@ -376,8 +419,8 @@ def test_space_from_space():
     a = Real(2.0, 30.0, transform="normalize")
-    for i in range(50):
-        check_limits(a.rvs(random_state=i), 2, 30)
+    # for i in range(50):
+    #     check_limits(a.rvs(random_state=i), 2, 30)
```
Why was this commented out, and not removed? Same comment for lines 438-439, 452-453, 479-480 and 494-495.

dk-teknologisk-mon commented on ProcessOptimizer/tests/test_space.py:
```diff
@@ -677,31 +707,42 @@ def test_purely_categorical_space():
+    num_min_pos = np.where(is_min_pos[0:3])[1]
+    # We only bother for the numerical axes, since the categorical axis has fewer than six values, so it has repeated values.
+    assert len(np.unique(num_min_pos)) > 1
+    # The minimum value should not always be in the same position
```
```suggestion
    # For a space of this size, two LHS samples with six points should always be different
```

dk-teknologisk-mon commented on ProcessOptimizer/tests/test_space.py:
```diff
@@ -677,31 +707,42 @@ def test_purely_categorical_space():
+    # Verfying that lhs can be performed randomly
+    for i in range(4):
+        assert set([x[i] for x in lhs_one]) == set([x[i] for x in lhs_two])
+    # Asserting the the values are the same for both the lhs, even though the order is different
```
```suggestion
    # Asserting the the values are the same for both the lhs, even though the order is different
    for i in range(4):
        assert set([x[i] for x in lhs_one]) == set([x[i] for x in lhs_two])    
```
Moving comment up above the code it refers to for consistency.

dk-teknologisk-mon commented on ProcessOptimizer/tests/test_space.py:
```diff
@@ -677,31 +707,42 @@ def test_purely_categorical_space():
+    values = [[sample[i] for sample in samples] for i in range(4)]
+    # set disregards order, so we use it to test which values we got.
+    assert set(values[0]) == set(range(1, 6 + 1))
+    # We should get all six integers, in any order
```
```suggestion
    # Set disregards order, so we use it to test which values we got.
    # We should get all six integers, in any order
    assert set(values[0]) == set(range(1, 6 + 1))
```
Moving comment up above the code it refers to for consistency.

dk-teknologisk-mon commented on ProcessOptimizer/tests/test_space.py:
```diff
@@ -677,31 +707,42 @@ def test_purely_categorical_space():
+    assert set(values[0]) == set(range(1, 6 + 1))
+    # We should get all six integers, in any order
+    assert set(values[1]) == set(0.5 + np.arange(1, 7))
+    # Six reals between 1 and 7, evenly divide, lie halfway between integers
```
```suggestion
    # Six reals between 1 and 7, evenly divide, lie halfway between integers
    assert set(values[1]) == set(0.5 + np.arange(1, 7))
```
Moving comment up above the code it refers to for consistency.

dk-teknologisk-mon commented on ProcessOptimizer/tests/test_space.py:
```diff
@@ -677,31 +707,42 @@ def test_purely_categorical_space():
+            for value in np.sqrt(10) * 10**-3 * 10 ** np.arange(6)
+        )
+    # there are no good tests for approximate equality that disregards order
+    # For log-uniform, we should get six values evenly spaced on a log scale
```
```suggestion
    # There are no good tests for approximate equality that disregards order
    # For log-uniform, we should get six values evenly spaced on a log scale
    for log_real in values[2]:
        assert any(
            log_real == pytest.approx(value)
            for value in np.sqrt(10) * 10**-3 * 10 ** np.arange(6)
        )
```
Moving comment up above the code it refers to for consistency.

dk-teknologisk-mon commented on ProcessOptimizer/tests/test_space.py:
```diff
@@ -677,31 +707,42 @@ def test_purely_categorical_space():
+    lhs_one = SPACE.lhs(n=6, random_state=None)
+    lhs_two = SPACE.lhs(n=6, random_state=None)
+    assert lhs_one != lhs_two
+    # Verfying that lhs can be performed randomly
```
```suggestion
    
```

dk-teknologisk-mon commented on ProcessOptimizer/utils/get_rng.py:
```diff
@@ -0,0 +1,35 @@
+        return np.random.default_rng(input)
+    elif isinstance(input, np.random.RandomState):
+        return np.random.default_rng(input.randint(1000, size=10))
+    # Draws 10 integers under 1000 from the deprecate RandomState to use as a seed for
```
```suggestion
    # Draws 10 integers under 1000 from the deprecated RandomState to use as a seed for
```

dk-teknologisk-mon commented on ProcessOptimizer/utils/utils.py:
```diff
@@ -1,886 +1,896 @@
+        if x not in space:
+            raise ValueError(
+                "Point (%s) is not within the bounds of"
+                " the space (%s)." % (x, space.bounds)
```
```suggestion
                "Point (%s) is not within the bounds of "
                "the space (%s)." % (x, space.bounds)
```

dk-teknologisk-mon commented on ProcessOptimizer/utils/utils.py:
```diff
@@ -1,886 +1,896 @@
+def check_x_in_space(x, space):
+    if is_2Dlistlike(x):
+        if not np.all([p in space for p in x]):
+            raise ValueError("Not all points are within the bounds of" " the space.")
```
```suggestion
            raise ValueError("Not all points are within the bounds of the space.")
```

dk-teknologisk-mon commented on ProcessOptimizer/utils/utils.py:
```diff
@@ -1,886 +1,896 @@
+        if not np.all([p in space for p in x]):
+            raise ValueError("Not all points are within the bounds of" " the space.")
+        if any([len(p) != len(space.dimensions) for p in x]):
+            raise ValueError("Not all points have the same dimensions as" " the space.")
```
```suggestion
            raise ValueError("Not all points have the same dimensions as the space.")
```

dk-teknologisk-mon commented on ProcessOptimizer/utils/utils.py:
```diff
@@ -1,886 +1,896 @@
+
+
+def expected_minimum(
+    res, n_random_starts=20, random_state=None, return_std=False, minmax="min"
```
```suggestion
    res, 
    n_random_starts=20,
    random_state=None,
    return_std=False,
    minmax="min",
```

dk-teknologisk-mon commented on ProcessOptimizer/utils/utils.py:
```diff
@@ -1,886 +1,896 @@
+            return -1 * reg.predict(x.reshape(1, -1))[0]
+        else:
+            raise ValueError(
+                "expected minmax to be in ['min','max'], got %s" % (minmax)
```
```suggestion
                "Expected minmax to be in ['min','max'], got %s" % (minmax)
```

dk-teknologisk-mon commented on ProcessOptimizer/utils/utils.py:
```diff
@@ -1,886 +1,896 @@
+            return [v for v in best_x], -best_fun
+
+    else:
+        raise ValueError("expected acq_func to be in ['min','max'], got %s" % (minmax))
```
```suggestion
        raise ValueError("Expected acq_func to be in ['min','max'], got %s" % (minmax))
```

dk-teknologisk-mon commented on ProcessOptimizer/utils/utils.py:
```diff
@@ -1,886 +1,896 @@
+
+def expected_minimum_random_sampling(
+    res, n_random_starts=100000, random_state=None, return_std=False, minmax="min"
+):
```
```suggestion
def expected_minimum_random_sampling(
    res,
    n_random_starts=100000,
    random_state=None,
    return_std=False,
    minmax="min",
):
```

dk-teknologisk-mon commented on ProcessOptimizer/utils/utils.py:
```diff
@@ -1,886 +1,896 @@
+    elif minmax == "max":
+        index_best_objective = np.argmax(y_random)
+    else:
+        raise ValueError("expected minmax to be in ['min','max'], got %s" % (minmax))
```
```suggestion
        raise ValueError("Expected minmax to be in ['min','max'], got %s" % (minmax))
```

dk-teknologisk-mon commented on ProcessOptimizer/utils/utils.py:
```diff
@@ -1,886 +1,896 @@
+
+def cook_estimator(
+    base_estimator, space=None, length_scale_bounds=None, length_scale=None, **kwargs
+):
```
```suggestion
def cook_estimator(
    base_estimator,
    space=None,
    length_scale_bounds=None,
    length_scale=None,
    **kwargs,
):
```

dk-teknologisk-mon commented on ProcessOptimizer/utils/utils.py:
```diff
@@ -1,886 +1,896 @@
+        if base_estimator not in ["GP", "ET", "RF", "GBRT", "DUMMY"]:
+            raise ValueError(
+                "Valid strings for the base_estimator parameter "
+                " are: 'RF', 'ET', 'GP', 'GBRT' or 'DUMMY' not "
```
```suggestion
                "are: 'RF', 'ET', 'GP', 'GBRT' or 'DUMMY' not "
```

dk-teknologisk-mon commented on ProcessOptimizer/utils/utils.py:
```diff
@@ -1,886 +1,896 @@
+    observed_max = res.func_vals.max()
+    min_x, expected_min = expected_minimum(
+        res, n_random_starts=20, random_state=None, minmax="min"
+    )
```
```suggestion
    min_x, expected_min = expected_minimum(res, random_state=random_state)
```
Remove calls of default parameters.

dk-teknologisk-mon commented on ProcessOptimizer/utils/utils.py:
```diff
@@ -1,886 +1,896 @@
+    )
+    max_x, expected_max = expected_minimum(
+        res, n_random_starts=20, random_state=None, minmax="max"
+    )
```
```suggestion
    max_x, expected_max = expected_minimum(res, random_state=random_state, minmax="max")
```
Remove calls of default parameters.

dk-teknologisk-mon commented on ProcessOptimizer/utils/utils.py:
```diff
@@ -1,886 +1,896 @@
+            [
+                max_x,
+            ]
+        )
```
```suggestion
        min_x = res.space.transform([min_x, ])
        max_x = res.space.transform([max_x, ])
```
Simple enough that keeping it on one line improves readability.

dk-teknologisk-mon committed:
Updates CHANGELOG.md

dk-teknologisk-mon committed:
Update CHANGELOG.md

Make it clear LHS is still seeded by default.

dk-teknologisk-mon committed:
Improves readability

dk-teknologisk-mon committed:
Applies suggestions from code review

Many small fixes to improve readability of the code. Also, I finally found the button to batch these commits...

dk-teknologisk-mon commented on ProcessOptimizer/space/space.py:
```diff
@@ -1,917 +1,941 @@
+
+    if len(dimension) == 3:
+        if any([isinstance(dim, (float, int)) for dim in dimension[:2]]) and 
+            dimension[2] in ["uniform", "log-uniform"]:
```
```suggestion
                dimension[2] in ["uniform", "log-uniform"]:
```

dk-teknologisk-mon reviewed (commented):
Fix of formatting in if-sentence

dk-teknologisk-mon committed:
Fix formatting in if-sentence

dk-teknologisk-mon commented on ProcessOptimizer/space/space.py:
```diff
@@ -1,917 +1,941 @@
+            )
+
+    if len(dimension) == 3:
+        if any([isinstance(dim, (float, int)) for dim in dimension[:2]]) and 
+                dimension[2] in ["uniform", "log-uniform"]:
```
```suggestion
        if (any([isinstance(dim, (float, int)) for dim in dimension[:2]]) and 
                dimension[2] in ["uniform", "log-uniform"]):
```

dk-teknologisk-mon committed:
Try at fixing if-statement

dk-teknologisk-mon commented on ProcessOptimizer/tests/test_space.py:
```diff
@@ -677,31 +707,42 @@ def test_purely_categorical_space():
+    # We only bother for the numerical axes, since the categorical axis has fewer than six values, so it has repeated values.
+    assert len(np.unique(num_min_pos)) > 1
+    # For a space of this size, two LHS samples with six points should always be different
+    lhs_one = SPACE.lhs(n=6, random_state=None)
+    lhs_two = SPACE.lhs(n=6, random_state=None)
```
```suggestion
    lhs_one = SPACE.lhs(n=6, seed=None)
    lhs_two = SPACE.lhs(n=6, seed=None)
```

dk-teknologisk-mon committed:
Updates test of LHS with new keyword name

dk-teknologisk-mon commented:
@SRFU-NN Will you pull the scikit-learn fix from develop into this branch? This should hopefully fix all the tests.

SRFU-NN commented:
> @SRFU-NN Will you pull the scikit-learn fix from develop into this branch? This should hopefully fix all the tests.

Will do.

SRFU-NN committed:
Merge remote-tracking branch 'origin/develop' into sampling

SRFU-NN committed:
Merge branch 'sampling' of https://github.com/novonordisk-research/ProcessOptimizer into sampling

SRFU-NN commented:
> @SRFU-NN Will you pull the scikit-learn fix from develop into this branch? This should hopefully fix all the tests.

@dk-teknologisk-mon : Tests pass. Can I merge the pull request, or is there more reviewing to be done?
